components/ntp_background_images/browser/ntp_background_images_service.cc

```diff
@@ -97,32 +421,73 @@ bool NTPBackgroundImagesService::HasObserver(Observer* observer) {
 
 NTPBackgroundImagesData*
 NTPBackgroundImagesService::GetBackgroundImagesData() const {
-  if (base::FeatureList::IsEnabled(features::kBraveNTPBrandedWallpaperDemo))
-    return GetDemoWallpaper();
-
   if (images_data_ && images_data_->IsValid())
     return images_data_.get();
 
   return nullptr;
 }
 
 void NTPBackgroundImagesService::OnComponentReady(
+    bool is_super_referrer,
     const base::FilePath& installed_dir) {
-  // image list is no longer valid after the component has been updated
-  images_data_.reset();
+  // Early return for test data because any components are not used when test
+  // data is used.
+  if (test_data_used_) {
+    GetComponentJsonData(installed_dir);
+    return;
+  }
+
+  // If we request component(1) registration after asking component(2), we can't
+  // guarantee component(1) is delivered first than 2.
+  // If (1) is delivered lately, we should ignore it and un register it. In this
+  // case, we should not update |images_data_| with it.
+  // Usually, we tries to unregister other type of component when a component
+  // is registered. However, both components could be in ready state when both
+  // are registered in a row. In this case, component(1) should be unregistered.
+  // One example is - we register NTP SI if we don't have cached referall code.
+  // When referral service fetched a code, we can register NTP SR if it's super
+  // referral code.
+  if (is_super_referrer != is_super_referrer_active_) {
+    const std::string code = GetCachedReferralPromoCode();
+    is_super_referrer ? UnRegisterSuperReferrerComponentIfRunning(code)
+                      : UnRegisterSponsoredImagesComponentIfRunning();
+    return;
+  }
+
+  GetComponentJsonData(installed_dir);
+}
+
+void NTPBackgroundImagesService::GetComponentJsonData(
+  const base::FilePath& installed_dir) {
   installed_dir_ = installed_dir;
   base::PostTaskAndReplyWithResult(
       FROM_HERE, {base::ThreadPool(), base::MayBlock()},
-      base::BindOnce(&ReadPhotosManifest,
-                     installed_dir.AppendASCII(kPhotoJsonFilename)),
-      base::BindOnce(&NTPBackgroundImagesService::OnGetPhotoJsonData,
+      base::BindOnce(&ReadComponentJsonData, installed_dir_),
+      base::BindOnce(&NTPBackgroundImagesService::OnGetComponentJsonData,
                      weak_factory_.GetWeakPtr()));
 }
 
-void NTPBackgroundImagesService::OnGetPhotoJsonData(
-    const std::string& photo_json) {
-  images_data_.reset(new NTPBackgroundImagesData(photo_json, installed_dir_));
+void NTPBackgroundImagesService::OnGetComponentJsonData(
+    const std::string& json_string) {
+  const bool was_super_referrer =
```

Comment:
why are we getting this from the images data and you also have is_super_referrer_active_ plus there's a pref for the referral promo code? I think you need one method that returns the correct component_id to load based on the prefs and remove these flags and other checks scattered around the code which won't work when there are more than two possible components anyway

Reply:
I thought judging with current data is most accurate and simple in here.
|is_super_referrer_active_| doesn't mean that current active component is SR.
It means SR component registration is asked lastly.
I think |is_super_referrer_active_| should be renamed like |is_super_referrer_lastly_asked_component_| to avoid confusing.